File: app/src/main/java/com/zq/opengl/opengl/widget/CameraRender.java
package com.zq.opengl.opengl.widget;

import android.content.Context;
import android.graphics.SurfaceTexture;
import android.media.MediaRecorder;
import android.opengl.EGL14;
import android.opengl.GLSurfaceView;
import android.os.Environment;

import androidx.camera.core.Preview;
import androidx.lifecycle.LifecycleOwner;


import com.zq.opengl.opengl.filter.AbstractFilter;
import com.zq.opengl.opengl.filter.BigEyeFilter;
import com.zq.opengl.opengl.filter.CameraFilter;
import com.zq.opengl.opengl.filter.FilterChain;
import com.zq.opengl.opengl.filter.FilterContext;
import com.zq.opengl.opengl.filter.ScreenFilter;
import com.zq.opengl.opengl.record.VideoEncodeRecode;
import com.zq.opengl.opengl.utils.CameraHelper;
import com.zq.opengl.opengl.utils.OpenGLUtils;

import java.io.File;
import java.util.ArrayList;
import java.util.List;


import javax.microedition.khronos.egl.EGLConfig;
import javax.microedition.khronos.opengles.GL10;

public class CameraRender implements GLSurfaceView.Renderer, Preview.OnPreviewOutputUpdateListener, SurfaceTexture.OnFrameAvailableListener {
    private CameraView cameraView;
    private CameraHelper cameraHelper;
    // 摄像头的图像  用OpenGL ES 画出来
    private SurfaceTexture mCameraTexure;

    private int[] textures;

    float[] mtx = new float[16];
    private MediaRecorder mRecorder;
    private FilterChain filterChain;

    private VideoEncodeRecode mVideoEncodeRecode;

    public CameraRender(CameraView cameraView) {
        this.cameraView = cameraView;
        OpenGLUtils.copyAssets2SdCard(cameraView.getContext(), "lbpcascade_frontalface.xml", "/sdcard/lbpcascade_frontalface.xml");
        OpenGLUtils.copyAssets2SdCard(cameraView.getContext(), "pd_2_00_pts5.dat", "/sdcard/pd_2_00_pts5.dat");
        LifecycleOwner lifecycleOwner = (LifecycleOwner) cameraView.getContext();
        cameraHelper = new CameraHelper(lifecycleOwner, this);
    }

    @Override
    public void onSurfaceCreated(GL10 gl, EGLConfig config) {
        //创建OpenGL 纹理 ,把摄像头的数据与这个纹理关联
        textures = new int[1];  //当做能在opengl用的一个图片的ID
        mCameraTexure.attachToGLContext(textures[0]);
        // 当摄像头数据有更新回调 onFrameAvailable
        mCameraTexure.setOnFrameAvailableListener(this);

        Context context = cameraView.getContext();

        List<AbstractFilter> filters = new ArrayList<>();
        filters.add(new CameraFilter(context));
        filters.add(new BigEyeFilter(context));
        filters.add(new ScreenFilter(context));

        filterChain = new FilterChain(filters, 0, new FilterContext());

        //录制视频的宽、高
        mVideoEncodeRecode = new VideoEncodeRecode(
                context, EGL14.eglGetCurrentContext(),
                Environment.getExternalStorageDirectory().getAbsolutePath() + File.separator +
                        "testRecode.mp4",
                480,
                640//相机的宽和高是相反的
        );
    }

    @Override
    public void onSurfaceChanged(GL10 gl, int width, int height) {

        filterChain.setSize(width, height);
    }

    @Override
    public void onDrawFrame(GL10 gl) {
        //todo 更新纹理
        mCameraTexure.updateTexImage();

        mCameraTexure.getTransformMatrix(mtx);

        filterChain.setTransformMatrix(mtx);
        filterChain.setFace(cameraHelper.getFace());
        int id = filterChain.proceed(textures[0]);

        mVideoEncodeRecode.fireFrame(id, mCameraTexure.getTimestamp());
    }

    public void onSurfaceDestroyed() {

        filterChain.release();
    }


    /**
     * 更新
     *
     * @param output 預覽輸出
     */
    @Override
    public void onUpdated(Preview.PreviewOutput output) {
        mCameraTexure = output.getSurfaceTexture();
    }

    @Override
    public void onFrameAvailable(SurfaceTexture surfaceTexture) {
        //  请求执行一次 onDrawFrame
        cameraView.requestRender();
    }

    public void startRecord(float speed) {
        try {
            mVideoEncodeRecode.startRecode();
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public void stopRecord() {
        mVideoEncodeRecode.stopRecode();
    }
}

File: app/src/main/java/com/zq/opengl/opengl/widget/CameraRender.kt
package com.zq.opengl.opengl.widget

import android.content.Context
import android.graphics.SurfaceTexture
import android.graphics.SurfaceTexture.OnFrameAvailableListener
import android.opengl.EGL14
import android.opengl.GLSurfaceView
import android.os.Environment
import androidx.camera.core.Preview.OnPreviewOutputUpdateListener
import androidx.camera.core.Preview.PreviewOutput
import androidx.lifecycle.LifecycleOwner
import com.zq.opengl.opengl.filter.CameraFilter
import com.zq.opengl.opengl.filter.ScreenFilter
import com.zq.opengl.opengl.record.BaseVideoEncoder
import com.zq.opengl.opengl.record.PutPcmThread
import com.zq.opengl.opengl.record.VideoEncodeRecode
import com.zq.opengl.opengl.utils.CameraHelper
import java.io.File
import java.io.IOException
import javax.microedition.khronos.egl.EGLConfig
import javax.microedition.khronos.opengles.GL10


/**
 * @program:
 * @description:
 * @author: 闫世豪
 * @create: 2021-03-17 10:18
 */
class CameraRender(private val cameraView: CameraView) : GLSurfaceView.Renderer,
    OnFrameAvailableListener, OnPreviewOutputUpdateListener {
    // 摄像头的图像  用OpenGL ES 画出来
    private lateinit var mCameraTexture: SurfaceTexture
    private lateinit var screenFilter: ScreenFilter
    private lateinit var textures: IntArray
    var mtx: FloatArray = FloatArray(16)
    private val context: Context = cameraView.context
    private lateinit var cameraFilter: CameraFilter
    private val lifecycleOwner: LifecycleOwner = cameraView.context as LifecycleOwner
    private lateinit var mVideoEncodeRecode: VideoEncodeRecode


    protected lateinit var putPcmThread: PutPcmThread

    public override fun onSurfaceCreated(gl: GL10, config: EGLConfig) {
        //创建OpenGL 纹理 ,把摄像头的数据与这个纹理关联
        textures = IntArray(1) //当做能在opengl用的一个图片的ID
        mCameraTexture.attachToGLContext(textures.get(0))
        // 当摄像头数据有更新回调 onFrameAvailable
        mCameraTexture.setOnFrameAvailableListener(this)
        cameraFilter = CameraFilter(context)
        screenFilter = ScreenFilter(context)

        mVideoEncodeRecode = VideoEncodeRecode(
            cameraView.context, EGL14.eglGetCurrentContext(),
            Environment.getExternalStorageDirectory().absolutePath + File.separator +
                    "testRecode.mp4",
            480,
            640//相机的宽和高是相反的
        )

//
//        mVideoEncodeRecode.setOnStatusChangeListener { status ->
//            if (status === BaseVideoEncoder.OnStatusChangeListener.STATUS.START) {
//                putPcmThread =
//                    PutPcmThread(context, mVideoEncodeRecode)
//                putPcmThread.start()
//            }
//        }

    }

    public override fun onSurfaceChanged(gl: GL10, width: Int, height: Int) {
        screenFilter.setSize(width, height)
        cameraFilter.setSize(width, height)
    }

    public override fun onDrawFrame(gl: GL10) {
        mCameraTexture.updateTexImage()
        mCameraTexture.getTransformMatrix(mtx)
        cameraFilter.setTransformMatrix(mtx)
        var id: Int = cameraFilter.onDraw(textures[0])
        id = screenFilter.onDraw(id)
        mVideoEncodeRecode.fireFrame(id, mCameraTexture.timestamp)
    }

    fun onSurfaceDestroyed() {
        cameraFilter.release()
        screenFilter.release()
    }

    public override fun onFrameAvailable(surfaceTexture: SurfaceTexture) {
        cameraView.requestRender()
    }

    public override fun onUpdated(output: PreviewOutput) {
        mCameraTexture = output.surfaceTexture
    }

    fun startRecord(speed: Float) {
        try {

            mVideoEncodeRecode.startRecode()
            //    mRecorder.start(speed)
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    fun stopRecord() {
        mVideoEncodeRecode.stopRecode()
     //   putPcmThread.setExit(true)

    }

    init {
        CameraHelper(lifecycleOwner, this)
    }


}